### ISSUE TITLE
[css-ui] Auto-disable native appearance when cascaded value has author origin

### ISSUE THREAD

**Issue by andruud** [w3c/csswg-drafts#4777]
Browsers currently automatically disable the native appearance of widgets in some cases, even if 
`appearance` is e.g. `auto` (see [issue 11](https://drafts.csswg.org/css-ui-4/#index)). This can happen when the author styles the `border`, `background`, or in Firefox's case `padding` (I believe).

This issue is not about _which_ properties disable the native appearance, but _how_ they disable appearance, which is quite different in Firefox (which has reasonable behavior) vs Chrome/Safari (which has ... less reasonable behavior):

In Chrome (and as far as I can see in Safari), the native appearance is disabled when the author styles the background/border in such a way that it differs from the _would-be_ background/border had only UA-origin declarations existed. This approach isn't great, because:

- Tweaking the UA-style due to e.g. a redesign of widgets can toggle native behavior on or off in the wild. (This exact thing happened in Chrome recently).
- You can get different results in multiple dimensions: for each type of widget, each OS, each browser, depending on what's specified in the UA sheets. This also makes it entirely incomprehensible for authors. 
- It adds annoying technical complexity, since we must somehow know the computed values of an alternate cascade-reality that shouldn't need to exist.

Instead, we should specify that border/background/etc applied by the author origin disables the native appearance _regardless of their value_. This is (to the best of my knowledge) what Firefox does currently.

See also related [Intent to Ship on blink-dev](https://groups.google.com/a/chromium.org/forum/?authuser=0#!topic/blink-dev/uNKeyd8091A).

**Comment by zcorpan** [comment]
https://github.com/whatwg/html/pull/4857 has a start for defining this. See https://whatpr.org/html/4857/rendering.html#appropriate-widget

https://github.com/whatwg/html/pull/4857#issuecomment-525261650 is not yet fixed, though.

**Comment by smfr** [comment]
> [whatwg/html#4857](https://github.com/whatwg/html/pull/4857) has a start for defining this. See https://whatpr.org/html/4857/rendering.html#appropriate-widget

Shouldn't CSS define this, rather than HTML?

**Comment by smfr** [comment]
A downside of the Gecko behavior is that an author can't "get back" to the UA rendering by setting the property value to one that matches the UA stylesheet.

**Comment by Loirooriol** [comment]
> an author can't "get back" to the UA rendering

Maybe [`revert`](https://drafts.csswg.org/css-cascade-4/#default) could do this?

**Comment by ExE-Boss** [comment]
**Firefox** already supports `revert` since version 67 ([bug 1215878](https://bugzilla.mozilla.org/show_bug.cgi?id=1215878)).

**Comment by Loirooriol** [comment]
@ExE-Boss But `<button style="border: revert">Button</button>` disables the native appearance in Firefox.

**Comment by emilio** [comment]
Yeah, I think we could support `revert` switching back to the native appearance pretty easily though.

**Comment by emilio** [comment]
@andruud looking at [the blink implementation](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/css/resolver/style_cascade.cc;l=837;drc=0363d1ddfe4a89830234571d9c3e602e8692c48f), it seems you account for border-image properties too, which is something that we don't do: Is that intentional?

I'm not quite sure if we do that intentionally or not (would need to dig the history a bit). Both ignoring them and not ignoring them makes some amount of sense, we probably just need to agree on which one makes more sense or is more useful to authors.

**Comment by andruud** [comment]
@emilio It's intentional in the sense that I (for now) didn't intend to change _which_ properties disable appearance, just _how_ it happens. 

I guess if other visual aspects of the border (color, width, etc) are grounds for disabling, then border-image* is as well? (I don't have a strong opinion). I'm in general not a very big fan of this auto-disabling feature, but since we have it, we should probably always auto-disable when an author declaration would effectively be ignored by maintaining the appearance.

`*` And maybe border-radius?

**Comment by zcorpan** [comment]
@andruud commented in https://github.com/whatwg/html/pull/4857#discussion_r435044451 that we might want to include animation origin and transition origin as well, not only author origin.

**Comment by zcorpan** [comment]
> > [whatwg/html#4857](https://github.com/whatwg/html/pull/4857) has a start for defining this. See https://whatpr.org/html/4857/rendering.html#appropriate-widget
> 
> Shouldn't CSS define this, rather than HTML?

Yes, probably some things should be moved to CSS UI. Also see https://github.com/w3c/csswg-drafts/issues/3526
